Docker-based QPY compatibility tests

gadial opened the pull request:
### Summary
Uses docker to build older qiskit versions for QPY compatibility testing.

* Addresses #15139


### Details and comments
The current QPY compatibility flow uses the same python version to install all the previous versions of qiskit, ignoring the ones not supported by the current python version. This would start causing problems when python 3.9 support is dropped, as described in #15139.

This PR changes the way old qiskit versions are generated by switching to a docker-based flow. For every qiskit version, its minimum supported python version is inferred from the `pypi` data, and an appropriate python base image is used for the docker build. The changes required are minimal, and the testing script itself is untouched.

gadial committed:
get_version now returns the minimum python version for each package

gadial committed:
Switching to docker-based QPY file generation for old versions

gadial committed:
Removes skipping versions not compatible with the current python version

gadial committed:
Linting

gadial committed:
Merge branch 'main' into docker_based_qpy_compat

eliarbel reviewed (commented):
I think it makes sense and would be useful to still have a way to run the compatibility tests locally without using docker.

gadial commented:
> I think it makes sense and would be useful to still have a way to run the compatibility tests locally without using docker.

No problem, I can return the relevant code and condition it on a `--no-docker` flag passed to `run_tests.sh`. Note however that dockerless local run is non comprehensive in the sense that some qiskit versions are not tested, since they are incompatible with the local python version. So working with docker might be the better way of doing this in practice.

gadial committed:
Returning the old, venv based mode as "--no-docker" mode

gadial commented:
Ok, added a `--no-docker` option. Trying to keep both paths in the same `process_version` file resulted in code I wasn't too happy with (because parameter handling became a pain), so I split in two files.

gadial committed:
Changed the way get_versions works to obtain the maximum python version for a package, not the minimum

gadial committed:
get_versions() should only use python versions which have concrete installable wheels

gadial committed:
Small fix for venv run, and debugging info for docker run since it runs out of space on the server

gadial committed:
Add docker build cache pruning to prevent out of space errors on the server

debasmita2102 commented on test/qpy_compat/process_version.sh:
```diff
@@ -31,30 +31,42 @@ while getopts "p:" opt; do
+    echo "Current docker space usage:"
+    docker system df
+    echo "System free space:"
+    df -h /
```
I was wondering about disk space management with this parallel Docker builds. In `process_version.sh` lines 63-69, you're running `docker builder prune -f` after each QPY file generation. Since `run_tests.sh` runs versions in parallel with parallel -j 2 (line 54), I'm wondering if Docker images could accumulate simultaneously before cleanup triggers.

Example: If Job 1 builds a ~600MB image AND Job 2 builds another ~600MB image at the same time (line 57), that's 1.2GB peak disk usage before either cleanup runs. On systems with limited storage, can this fail before docker builder prune recovers space. I might be overthinking this, but if this is a concern, should there be a final cleanup phase at the end of run_tests.sh (after all parallel jobs complete) to ensure any remaining dangling images are purged? I'm curious how you'd approach this.

jakelishman replied:
Debasmita: in theory we already had this problem with existing format, since the venvs were also constructed and only removed at the end of the `process_version.sh` script.  Since Docker's cleaning up its image after `docker run` (because of the `--rm` flag), I don't _think_ (but am pretty unsure) that the situation is hugely worse that it was before, and we hadn't yet hit any storage problems in CI.  1.2GB isn't super huge - I think building a single instance of Qiskit alone in debug mode already needs rather more storage than that in cargo's `target` directory, and in this part of the process, we're only installing things from binary artifacts.

gadial replied:
This is a valid concern which may be addressed by making the script carefully check the available memory and limit parallel executions accordingly. In practice I don't think this will pose a real problem, at least in the CI/CD flow (if I recall correctly space begins to be a problem only after 10 or so runs).

jakelishman commented on test/qpy_compat/run_tests.sh:
```diff
@@ -11,6 +11,11 @@
 # Any modifications or derivative works of this code must retain this
 # copyright notice, and modified files need to carry a notice indicating
 # that they have been altered from the originals.
+no_docker=false
+if [[ " $@ " =~ " --no-docker " ]]; then
+    no_docker=true
+fi
+
```
Minor: this should probably come below the `set` options - those define the settings for the script execution, and this block is part of the script.

Also, `false` and `true` are not variables at all, but actually executable programs that return exit codes 1 and 0 respectively.

gadial replied:
yeah, I was worried that using `true` and `false` is a little *too* hack-y

jakelishman commented on test/qpy_compat/get_versions.py:
```diff
@@ -63,36 +64,61 @@ def available_versions_for_package(package, min_version=None, max_version=None):
+                    ]  # convert "cp311" to "3.11"
+                    python_version = max(
+                        python_versions, key=lambda s: tuple(map(int, s.split(".")))
+                    )
```
Interestingly, I think this is going to cause us to do something like the non-monotonic:

- terra 0.18: Python 3.9
- terra 0.19--0.23: Python 3.10
- terra 0.24: Python 3.11
- terra 0.25--qiskit 1.2: Python 3.8
- qiskit 1.3--qiskit 2.2: Python 3.9
- qiskit 2.3+: Python 3.10

because we swapped from per-version wheels to abi3 wheels in terra 0.25, which means from that point on, the only listed Python version in a wheel tag will be the _oldest_ supported version.

I don't think that matters much, it's just interesting.

jakelishman commented on test/qpy_compat/get_versions.py:
```diff
@@ -63,36 +64,61 @@ def available_versions_for_package(package, min_version=None, max_version=None):
-                )
-                continue
-            yield other_version
+            if "--no-docker" in sys.argv:
```
(No need to make this change.)

Minor: the `--no-docker` flag might be better called `--any-python` or something like that - the behaviour is to do with the allowed versions of Python, rather than specifically anything to do with Docker.

gadial replied:
I don't agree here, since the flag is the user's way of saying whether he wants to run the tests using docker images or whether he wants everything built locally (which excludes some python versions, based on the user's environment).

jakelishman replied:
I agree with `--no-docker` for the user-facing drive `test/qpy_compat/run_tests.sh` for the reasons you say, but not for the inner worker `get_versions.py`, for the reasons I said.

jakelishman commented on test/qpy_compat/run_tests.sh:
```diff
@@ -42,7 +48,11 @@ python -m venv "$qiskit_venv"
 "$qiskit_venv/bin/pip" install -c "$repo_root/constraints.txt" "$qiskit_dev_wheel" packaging "symengine<0.14" "sympy>1.3"
 
 # Run all of the tests of cross-Qiskit-version compatibility.
-"$qiskit_python" "$our_dir/get_versions.py" | parallel -j 2 --colsep=" " bash "$our_dir/process_version.sh" -p "$qiskit_python"
+if $no_docker; then
+    "$qiskit_python" "$our_dir/get_versions.py" --no-docker | parallel -j 2 --colsep=" " bash "$our_dir/process_version_with_venv.sh" -p "$qiskit_python"
+else
+    "$qiskit_python" "$our_dir/get_versions.py" | parallel -j 2 --colsep=" " bash "$our_dir/process_version.sh" -p "$qiskit_python"
+fi
```
(No need to make this change.)

Maybe a slightly cleaner way of showing that these are the same thing would be
```bash
if [[ " $@ " =~ " --no-docker " ]]; then
  version_flags=""
  process_script="process_version_with_venv.sh"
else
  version_flags="--any-python"
  process_script="process_version.sh"
fi

"$qiskit_python" "$our_dir/get_versions.py" $version_flags | parallel -j 2 --colsep= " " bash "$our_dir/$process_script" -p "$qiskit_python"
```
since we're already doing the `if [[` check?  That way it's easier to modify the `parallel` arguments / whatever if we need to?

jakelishman commented on test/qpy_compat/dockerfile:
```diff
@@ -0,0 +1,12 @@
+ENV PIP_NO_CACHE_DIR=1 PIP_DISABLE_PIP_VERSION_CHECK=1 PYTHONDONTWRITEBYTECODE=1
+WORKDIR /work
+COPY qpy_test_constraints.txt .
+RUN python -m pip install --upgrade pip && pip install -c qpy_test_constraints.txt "${PACKAGE_NAME}==${PACKAGE_VERSION}"; 
```
I was vaguely worried that this might attempt to build Numpy and/or Scipy from source for older Pythons, but happily it doesn't seem to be the case, and `pip` is preferring binary only.

jakelishman commented on test/qpy_compat/process_version.sh:
```diff
@@ -31,30 +31,42 @@ while getopts "p:" opt; do
+    echo "Current docker space usage:"
+    docker system df
+    echo "System free space:"
+    df -h /
```
The space-printing things here might not be necessary any more and diluting the already polluted output of this script?  If we get killed because of storage, I would _think_ that GitHub would tell us that that was the reason.

gadial replied:
You can see [here](https://github.com/Qiskit/qiskit/actions/runs/18900534585/job/53946580854) how the "out of space" issue looked:

```
4.633 Downloading symengine-0.13.0-cp39-cp39-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (49.7 MB)
5.083 ERROR: Could not install packages due to an OSError: [Errno 28] No space left on device
5.083 
5.085    ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━        41.4/49.7 MB 92.8 MB/s eta 0:00:01
------
dockerfile:10
--------------------
   8 |     WORKDIR /work
   9 |     COPY qpy_test_constraints.txt .
  10 | >>> RUN python -m pip install --upgrade pip && pip install -c qpy_test_constraints.txt "${PACKAGE_NAME}==${PACKAGE_VERSION}"; 
  11 |     
  12 |     CMD ["python", "-c", "import qiskit; import sys; print(f'Qiskit version: {qiskit.__version__}'); print(f'Python version: {sys.version}')"]
--------------------
ERROR: failed to build: failed to solve: process "/bin/sh -c python -m pip install --upgrade pip && pip install -c qpy_test_constraints.txt \"${PACKAGE_NAME}==${PACKAGE_VERSION}\";" did not complete successfully: exit code: 1
parallel: Error: Output is incomplete.
parallel: Error: Cannot append to buffer file in /tmp.
parallel: Error: Is the disk full?
parallel: Error: Change $TMPDIR with --tmpdir or use --compress.
Error: Process completed with exit code 255.
```

so basically - yeah, no need for our prints to tell us this is the source of the problem, and if you don't think that's useful input we may as well remove it.

jakelishman replied:
I think we'd be ok to remove them now, just to slightly reduce the amount of output.  I guess that failure _could_ have bee clearer, but the suggestion in `parallel` is correct anyway, so if we _do_ have to debug it again, it should be clear what it is that's going wrong.

jakelishman reviewed (dismissed):
Thanks for doing this Gadi, it's really important work.

I left a small number of comments, most of which are just comments.  A couple of things I made suggestions, but I don't actually want to merge them right now anyway because I want this in ASAP and I don't want to go through more GitHub Actions debugging cycles right before the rc1.

The only real question I had was about whether we could/should remove the all the disk-storage printing stuff, but it's not critical.

I've kicked off a run in #15371 using a roll-up of this PR to test that everything will work correctly once Python 3.9 is formally dropped - I'll hold off adding this to the queue until we've got a response from that PR, in case you wanted to make any of the minor changes I mentioned.

gadial committed:
Remove the space-related debug prints

jakelishman reviewed (approved):
My trial run with Python 3.9 dropped has now passed all relevant CI (https://github.com/Qiskit/qiskit/actions/runs/19675026820/job/56353669438?pr=15371), so I think this is fine to go as-is, and anything left we want to change we can follow-up on, if we choose.

Thanks for getting this done, Gadi - it's a really vital part of our infrastructure.
